cjdns-core/src/encoding.rs

```diff
@@ -31,393 +32,614 @@
 //! assert_eq!(deserialized, input);
 //! ```
 
-use std::collections::HashSet;
-
-use thiserror::Error;
-
-use crate::EncodingSchemeForm;
-
-#[derive(Error, Debug, PartialEq, Eq)]
-pub enum EncodingSchemeError {
-    #[error("Invalid encoding scheme: no encoding forms defined")]
-    NoEncodingForms,
-    #[error("Invalid encoding scheme: too many encoding forms defined (max 31)")]
-    TooManyEncodingForms,
-    #[error("Invalid encoding scheme: single form has non-empty prefix")]
-    SingleFormWithPrefix,
-    #[error("Invalid encoding scheme: form has bit_count out of bounds (1..31)")]
-    BadBitCount,
-    #[error("Invalid encoding scheme: multiple forms - prefix length is out of bounds (1..31)")]
-    MultiFormBadPrefix,
-    #[error("Invalid encoding scheme: multiple forms should have bit_count in ascending order")]
-    BitCountNotSorted,
-    #[error("Invalid encoding scheme: multiple forms must have unique prefixes")]
-    DuplicatePrefix,
-    #[error("Invalid encoding scheme: form size too big (bit_count + prefix_len > 59)")]
-    TooBigForm,
-    #[error("Invalid serialized encoding scheme")]
-    BadSerializedData,
-    #[error("Invalid encoding form")]
-    BadEncodingForm,
-}
+pub use encoding_serde::{deserialize_forms, serialize_forms};
+pub use encoding_scheme::*;
+pub use errors::{SchemeValidationError, EncodingSerDeError};
+
+mod encoding_serde {
+    //! Serialization and deserialization logic
+
+    use super::EncodingSerDeError;
+    use crate::EncodingSchemeForm;
+
+    /// Store encoding scheme (array of `EncodingSchemeForm`) into a byte vector array (bits sequence).
+    ///
+    /// Accepts vector of `EncodingSchemeForm`s, encodes them as bits sequence
+    /// and returns the result as bytes vector.
+    pub fn serialize_forms(forms: &[EncodingSchemeForm]) -> Result<Vec<u8>, EncodingSerDeError> {
```

Comment:
Is this function used to encode any sequence of forms, or only a valid encoding scheme? In the latter case it it more practical to make this fn a method of `EncodingScheme`.